### DISCUSSION TITLE
"I'm lucky" merging feature

### DISCUSSION THREAD

**Discussion by miasik** [hako-mikan/sd-webui-supermerger#139] [Ideas]
There are two variants of my idea.

The first one is the simplest. It is only one button which randomizes ratio for each element and then does merge and render.

The second variant has three input boxes:

1. Minimum value
2. Maximum value
3. Runs amount

As in the first variant, the button randomizes ratios for the elements but using the corresponding values as min and max.

The second one can have some options:

Per block min/max values and randomization
Per block min/max values, but randomization is applied to each individual element of the block

**Comment by hako-mikan** [comment]
Hey there. Thanks for the suggestion. That feature is one I had been meaning to create but forgot about. I'll give it a try and create it.

**Reply by miasik** [reply]
Well, it also can be an option for XYZ.

**Comment by hako-mikan** [comment]
Now you can test it in dev.

**Comment by hako-mikan** [comment]
https://github.com/hako-mikan/sd-webui-supermerger/blob/dev/README.md#random-merge

**Reply by miasik** [reply]
I wanted to write that I finally managed how to work with it but there's still a mistake(or two) for me ;-)

XYZ always ignores my number for Y and produces only one iteration
Where can I see my randomly generated recipes?

For the next run I had 5 for Y and 3 for "Num of challenge". Anyway I got only one run

> 4 949375539, 2833106063, 949375541, 949375544, 949375545, 949375547, 949375548, 949375549 19 5 ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER,ER 0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2
XY plot start, mode:Add difference:A+(B-C)*alpha, X: seed, Y: random, Z: none MBW: True
xmen:949375539, 2833106063, 949375541, 949375544, 949375545, 949375547, 949375548, 949375549, ymen:5,zmen:, xtype:seed, ytype:random, ztype:none        weights_a:0, weights_b:0, model_A:Anime\Yoh-train-04-Hass13-66068.fp16.safetensors [5e0c20cc85],model_B :Photo\eroticVision_v3.safetensors [0981b250eb], model_C:Anime\Yoh-train-04-Hass13-66068.fp16.safetensors [5e0c20cc85], alpha:0.2,        beta :0.25, mode:Add difference:A+(B-C)*alpha, blocks:True
['949375539', '2833106063', '949375541', '949375544', '949375545', '949375547', '949375548', '949375549']
the die was thrown : ['949375539', '2833106063', '949375541', '949375544', '949375545', '949375547', '949375548', '949375549']
['']
XY plot: X: seed, 949375539, Y: random, , Z: none,  (1/3.0)
  model A       : Anime_Yoh-train-04-Hass13-66068.fp16
  model B       : Photo_eroticVision_v3
  model C       : Anime_Yoh-train-04-Hass13-66068.fp16
  alpha,beta    : (0.0, 0.25)
  weights_alpha : [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  weights_beta  : 0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2,0.2
  mode          : Add difference:A+(B-C)*alpha
  MBW           : True
  CalcMode      : trainDifference
  Elemental     : ['BASE IN00 IN01 IN02 IN03 IN04 IN05 IN06 IN07 IN08 IN09 IN10 IN11 M00 OUT00 OUT01 OUT02 OUT03 OUT04 OUT05 OUT06 OUT07 OUT08 OUT09 OUT10 OUT11::R']
  Weights Seed  : 3434675461
Loading weights [Photo_eroticVision_v3] from cache
Loading weights [Anime_Yoh-train-04-Hass13-66068.fp16] from cache
Loading weights [Anime_Yoh-train-04-Hass13-66068.fp16] from cache
Stage 1/2: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1143/1143 [00:19<00:00, 58.30it/s]
Stage 2/2: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1143/1143 [00:00<?, ?it/s]
Applying attention optimization: sdp-no-mem... done.
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.15it/s]
XY plot: X: seed, 2833106063, Y: random, , Z: none,  (2/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.51it/s]
XY plot: X: seed, 949375541, Y: random, , Z: none,  (3/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.52it/s]
XY plot: X: seed, 949375544, Y: random, , Z: none,  (4/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.51it/s]
XY plot: X: seed, 949375545, Y: random, , Z: none,  (5/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.52it/s]
XY plot: X: seed, 949375547, Y: random, , Z: none,  (6/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.52it/s]
XY plot: X: seed, 949375548, Y: random, , Z: none,  (7/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.52it/s]
XY plot: X: seed, 949375549, Y: random, , Z: none,  (8/3.0)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:04<00:00,  4.52it/s]
stopped at x=8,y=180it [47:37,  4.51it/s]

**Reply by miasik** [reply]
"set to alpha" and "set to beta" got broken and produce red Error in the interface

**Reply by hako-mikan** [reply]
Problem in XYZ plot was fixed.
You can see your ratios by enabling `print change` in XYZ plot options for `E` mode.
Or, check `save E-list`.

Tell me "set to alpha" and "set to beta" problem, in my environment, there is no errors.

**Reply by miasik** [reply]
As i see XY with random works well. I got a grid and the log wasn't buggy. I also saw saved recipes.

Here is the information related to "set to alpha". I pressed the button after a successful "XY random" run.
![изображение](https://github.com/hako-mikan/sd-webui-supermerger/assets/1073832/6a8fb802-c1c9-414a-b0f1-a75eb0ad204a)

> Traceback (most recent call last):
  File "e:\SD\automatic1111\venv\lib\site-packages\gradio\routes.py", line 422, in run_predict
    output = await app.get_blocks().process_api(
  File "e:\SD\automatic1111\venv\lib\site-packages\gradio\blocks.py", line 1323, in process_api
    result = await self.call_function(
  File "e:\SD\automatic1111\venv\lib\site-packages\gradio\blocks.py", line 1051, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "e:\SD\automatic1111\venv\lib\site-packages\anyio\to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "e:\SD\automatic1111\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "e:\SD\automatic1111\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "E:\SD\automatic1111\extensions\sd-webui-supermerger\scripts\supermerger.py", line 623, in slider2text
    if any(element in w for element in RANCHA):return w
  File "E:\SD\automatic1111\extensions\sd-webui-supermerger\scripts\supermerger.py", line 623, in <genexpr>
    if any(element in w for element in RANCHA):return w
TypeError: argument of type 'NoneType' is not iterable
